### ISSUE TITLE
Step::steps_between does not distinguish overflow and unimplemented (unstable)

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#48117]
It returns `Option<size>`: https://doc.rust-lang.org/std/iter/trait.Step.html#tymethod.steps_between

And the comment says it uses `None` for overflow: https://doc.rust-lang.org/src/core/iter/range.rs.html#29

But it also uses `None` for unimplemented https://doc.rust-lang.org/src/core/iter/range.rs.html#155

And as a result, `Range` needs to return `(0, None)` when it gets `None`: https://doc.rust-lang.org/src/core/iter/range.rs.html#235-240

It would be nice to either
- Require that the method is implemented accurately (easy for fundamental integers; harder if `Step` is expected to support graph walks, but maybe not harder than `PartialOrd`)
- Change the type to something else, maybe `Option<Option<usize>>` or be more hint-like with `(usize, Option<usize>)`
- Something else

**Comment by CAD97** [comment]
The links appear to have lost their original position.

Does the original concern still stand? (Does it stand with #69659?)

**Comment by scottmcm** [comment]
@CAD97 Yes, I think there's still essentially the same problem -- the Invariants section still has `None` for both unreachable (`steps_between(1, 0)`) and overflow (`steps_between(0_u128, 1<< 127)`).

So if this is used for `size_hint`, then getting `None` back from `steps_between` means you're stuck returning `(0, None)`, when you'd rather be able to return either `(0, Some(0))` or `(usize::MAX, None)`.

**Comment by scottmcm** [comment]
Looks like this was updated, but I realized it's unclear, at least to me, what to do if it's actually *unreachable* -- should that be marked distinctly?  Step only requires PartialOrd, not Ord, so `a` and `b` might both have successors and predecessors but neither is reachable from the other.

Or if `(0, None)` is always a legal response, shouldn't it be a provided implementation like `Iterator::step_by`?
